src/Form/Type/TpayRetryPaymentDetailsType.php

```diff
@@ -0,0 +1,159 @@
+                'blik_token',
+                TextType::class,
+                [
+                    'data' => null,
+                    'property_path' => '[blik_token]',
+                    'label' => 'commerce_weavers_sylius_tpay.shop.order_summary.blik.token',
+                    'validation_groups' => function (FormInterface $form): array {
+                        return [$this->getValidationGroup($form)];
+                    },
+                    'constraints' => [
+                        new NotBlank(groups: ['payment_retry_' . BlikGatewayFactory::NAME]),
+                        new Length(exactly: 6, groups: ['payment_retry_' . BlikGatewayFactory::NAME]),
+                    ],
+                ],
+            )
+            ->add(
+                'google_pay_token',
+                HiddenType::class,
+                [
+                    'property_path' => '[google_pay_token]',
+                    'label' => false,
+                    'validation_groups' => function (FormInterface $form): array {
+                        return [$this->getValidationGroup($form)];
+                    },
+                    'constraints' => [
+                        new EncodedGooglePayToken(groups: ['payment_retry_', GooglePayGatewayFactory::NAME]),
+                    ],
+                ],
+            )
+            ->add(
+                'apple_pay_token',
+                HiddenType::class,
+                [
+                    'property_path' => '[apple_pay_token]',
+                    'label' => false,
+                    'validation_groups' => ['sylius_checkout_complete'],
+                ],
+            )
+            ->add(
+                'tpay_channel_id',
+                HiddenType::class,
+                [
+                    'error_bubbling' => false,
+                    'property_path' => '[tpay_channel_id]',
+                    'validation_groups' => function (FormInterface $form): array {
+                        return [$this->getValidationGroup($form)];
+                    },
+                    'constraints' => [
+                        new ValidTpayChannel(groups: ['payment_retry_' . PBLGatewayFactory::NAME]),
+                    ],
+                ],
+            )
+            ->add(
+                'visa_mobile_phone_number',
+                TelType::class,
+                [
+                    'property_path' => '[visa_mobile_phone_number]',
+                    'attr' => [
+                        'placeholder' => 'commerce_weavers_sylius_tpay.shop.order_summary.visa_mobile.placeholder',
+                        'maxLength' => 15,
+                    ],
+                    'validation_groups' => function (FormInterface $form): array {
+                        return [$this->getValidationGroup($form)];
+                    },
+                    'constraints' => [
+                        new Length(min: 7, max: 15, groups: ['payment_retry_' . VisaMobileGatewayFactory::NAME]),
```

Comment:
we could add some function to convert it
```suggestion
                        new Length(min: 7, max: 15, groups: [$this->provideRequiredValidationGroup(VisaMobileGatewayFactory::NAME)]),
```

inside we can take `payment_retry_` from `self::VALIDATION_GROUP_PREFIX` const.